user_sync/connector/directory_okta.py

```diff
@@ -0,0 +1,290 @@
+
+        # self.user_email_formatter = OKTAValueFormatter(options['user_email_format'])
+
+        self.users_client = None
+        self.groups_client = None
+        self.logger = logger = user_sync.connector.helper.create_logger(options)
+
+        self.user_identity_type = user_sync.identity_type.parse_identity_type(options['user_identity_type'])
+
+        self.options = options
+        self.user_by_uid = {}
+
+        logger.debug('Initialized with options: %s', options)
+
+        logger.info('Connecting to: %s', okta_url)
+
+        try:
+            self.users_client = okta.UsersClient(okta_url, api_token)
+            self.groups_client = okta.UserGroupsClient(okta_url, api_token)
+        except Exception as e:
+            raise user_sync.error.AssertionException(repr(e))
+
+        logger.info('Connected')
+
+    def load_users_and_groups(self, groups, extended_attributes):
+        """
+        :type groups: list(str)
+        :type extended_attributes: list(str)
+        :rtype (bool, iterable(dict))
+        """
+
+        options = self.options
+
+        all_users_filter = options['all_users_filter']
+
+        is_using_source_filter = False
+        source_filters = options['source_filters']
+        source_filter = source_filters.get('all_users_filter')
+        if source_filter:
+            users_filter = "(%s) and (%s)" % (all_users_filter, source_filter)
+            is_using_source_filter = True
+            self.logger.info('Applied source filter: %s', users_filter)
+        else:
+            users_filter = all_users_filter
+
+        self.logger.info('Loading users...')
+        self.user_by_uid = user_by_uid = {}
+
+        for group in groups:
+            total_group_members = 0
+            total_group_users = 0
+            for user in self.iter_group_members(group, users_filter, extended_attributes):
+                total_group_members += 1
+
+                uid = user.get('uid')
+                if user and uid:
+                    if uid not in user_by_uid:
+                        user_by_uid[uid] = user
+                    total_group_users += 1
+                    user_groups = user_by_uid[uid]['groups']
+                    if group not in user_groups:
+                        user_groups.append(group)
+
+            self.logger.debug('Group %s members: %d users: %d', group, total_group_members, total_group_users)
+
+        return user_by_uid.itervalues()
+
+    def find_group(self, group):
+        """
+        :type group: str
+        :rtype UserGroup
+        """
+        group = group.strip()
+        options = self.options
+        group_filter_format = options['group_filter_format']
+        try:
+            results = self.groups_client.get_groups(query=group_filter_format.format(group=group))
+        except Exception as e:
+            self.logger.warning("Unable to query group")
+            raise user_sync.error.AssertionException(repr(e))
+
+        if results is None:
```

Comment:
No you always want to have an exception wrapper whenever you make a network call.  I just meant this should be a warning rather than a dead stop.  So this looks good.